File: src/eQuantic.UI.Components/SegmentedControl.cs
using eQuantic.UI.Primitives;

namespace eQuantic.UI.Components;

/// <summary>
/// A single choice out of two to five, all of them VISIBLE (spec B9) — the difference from a Select,
/// which hides its options until asked. Use it when the options are short, fixed and worth comparing
/// at a glance ("All · Income · Expenses"); past five segments the labels stop fitting and the
/// control becomes a Select. CONTROLLED: the caller owns <see cref="SelectedIndex"/>.
/// <para>
/// The selected segment is a raised thumb over a recessed track — the same Surface/SurfaceSubtle
/// pairing a Card uses, so selection reads as elevation rather than as a colour the user must learn.
/// Segments share the width equally, so the thumb never resizes as the choice moves.
/// </para>
/// </summary>
public sealed class SegmentedControl : StatelessComponent
{
    public SegmentedControl(IReadOnlyList<string> segments, int selectedIndex,
        Action<int>? onChanged = null)
    {
        Segments = segments;
        SelectedIndex = selectedIndex;
        OnChanged = onChanged;
    }

    public IReadOnlyList<string> Segments { get; init; }
    public int SelectedIndex { get; init; }
    public Action<int>? OnChanged { get; init; }

    public SizeVariant Size { get; init; } = SizeVariant.Medium;
    public bool Disabled { get; init; }

    /// <summary>Fills the available width; otherwise the control hugs its widest segment.</summary>
    public bool Stretch { get; init; } = true;

    public override VisualNode Build(ComponentContext context)
    {
        var theme = context.Theme;
        var height = Sizing.Height(Size);
        var inset = 3f;                                   // the track's lip around the thumb
        var trackRadius = Sizing.Radius(Size);

        var row = new Row(gap: 0) { Height = SizeValue.Fill, Cross = CrossAlign.Stretch };
        for (var i = 0; i < Segments.Count; i++)
        {
            var index = i;
            var selected = index == SelectedIndex;

            var label = new Row(gap: 0) { Main = MainAlign.Center, Cross = CrossAlign.Center };
            label.Add(new Text(Segments[index], TypeRole.Label,
                selected ? theme.TextPrimary : theme.TextSecondary, maxLines: 1)
            {
                StyleOverride = theme.Type(TypeRole.Label) with { Size = Sizing.LabelSize(Size) },
                Transition = TransitionSpec.Of(StyleChannels.Colors, Motion.Press),
            });

            var segment = new Box(new BoxStyle
            {
                Width = SizeValue.Fill,
                Height = SizeValue.Fill,
                Background = selected ? theme.Surface : null,
                CornerRadius = new CornerRadii(trackRadius - inset),
                Elevation = selected ? 1 : 0,
                Transition = TransitionSpec.Of(StyleChannels.Colors | StyleChannels.Shadow, Motion.Press),
            }, label);

            var press = new Pressable(segment, Disabled || selected ? null : () => OnChanged?.Invoke(index))
            {
                Disabled = Disabled,
                Label = Segments[index],
                Selected = selected,
            };

            row.Add(Stretch ? new Flexible(press, 1) : press);
        }

        return new Box(new BoxStyle
        {
            Width = Stretch ? SizeValue.Fill : SizeValue.Hug,
            Height = height,
            Padding = EdgeInsets.All(inset),
            Background = theme.SurfaceSubtle,
            CornerRadius = new CornerRadii(trackRadius),
            Opacity = Disabled ? theme.DisabledOpacity : 1f,
        }, row);
    }
}

File: src/eQuantic.UI.Runtime/src/shared/__transpiled__/SegmentedControl.ts
import { Adjustable, Box, BoxStyle, BuildContext, CornerRadii, EdgeInsets, Flexible, Motion, Pressable, Row, SizeValue, Sizing, StatelessComponent, Text, TransitionSpec } from "@equantic/runtime";

export class SegmentedControl extends StatelessComponent {
    declare segments: string[];
    declare selectedIndex: number;
    declare onChanged: any;
    declare size: string;
    declare disabled: boolean;
    declare stretch: boolean;
    constructor(segments?: any, selectedIndex?: any, onChanged: any = null, props?: any) {
        super();
        if (segments !== undefined) this.segments = segments;
        if (selectedIndex !== undefined) this.selectedIndex = selectedIndex;
        if (onChanged !== undefined) this.onChanged = onChanged;
        if (this.selectedIndex === undefined) this.selectedIndex = 0;
        if (this.size === undefined) this.size = 'medium';
        if (this.disabled === undefined) this.disabled = false;
        if (this.stretch === undefined) this.stretch = true;
        this.segments = segments;this.selectedIndex = selectedIndex;this.onChanged = onChanged;
        if (props && typeof props === 'object') Object.assign(this, props);
    }

    build(context: BuildContext) {
        let theme = context.theme;let height = Sizing.height(this.size, context.density);let inset = 3;let trackRadius = Sizing.radius(this.size);let row = new Row(0, 'start', 'center', false, null, null, { width: this.stretch ? SizeValue.fill : SizeValue.hug, height: SizeValue.fill, cross: 'stretch' });for (let i = 0; i < this.segments.length; i++) {let index = i;let selected = index === this.selectedIndex;let label = new Row(0, 'start', 'center', false, null, null, { width: this.stretch ? SizeValue.fill : SizeValue.hug, height: SizeValue.fill, main: 'center', cross: 'center' });label.add(new Text(this.segments[index], 'label', selected ? theme.textPrimary : theme.textSecondary, 1, 'start', false, false, null, { styleOverride: theme.type('label').withSize(Sizing.labelSize(this.size, context.density)), transition: TransitionSpec.of(1, Motion.press) }));let segment = new Box(new BoxStyle({ width: this.stretch ? SizeValue.fill : SizeValue.hug, height: SizeValue.fill, padding: this.stretch ? undefined : EdgeInsets.symmetric(Sizing.paddingX(this.size, context.density), 0), background: selected ? theme.surface : null, cornerRadius: new CornerRadii(trackRadius - inset), elevation: selected ? 1 : 0, transition: TransitionSpec.of(1 | 8, Motion.press) }), label);let press = new Pressable(segment, this.disabled || selected ? null : () => this.onChanged?.(index), { disabled: this.disabled, label: this.segments[index], selected: selected });row.add(this.stretch ? new Flexible(press, 1) : press);}let track = new Box(new BoxStyle({ width: this.stretch ? SizeValue.fill : SizeValue.hug, height: height, padding: EdgeInsets.all(inset), background: theme.surfaceSubtle, cornerRadius: new CornerRadii(trackRadius), opacity: this.disabled ? theme.disabledOpacity : 1 }), row);if (this.disabled || this.onChanged == null || this.segments.length === 0) return track;let count = this.segments.length;return new Adjustable(track, (direction: number) => this.onChanged((this.selectedIndex + direction + count) % count), { role: 'radiogroup' });
    }

}

